src/ocean/util/container/TreeMap.d

```diff
@@ -0,0 +1,599 @@
+
+struct TreeMap ( Node = eb64_node )
+{
+    import ocean.transition;
+
+    static if (!is(Node == eb64_node))
+    {
+        static assert(is(Node == struct));
+        static assert(is(typeof(Node.tupleof[0]) == eb64_node));
+        static assert(!Node.tupleof[0].offsetof);
+    }
+
+    /***************************************************************************
+
+        The `malloc` allocator.
+
+    ***************************************************************************/
+
+    private static MallocAllocator allocator;
+
+    /***************************************************************************
+
+        The ebtree root.
+
+    ***************************************************************************/
+
+    private eb_root root = empty_unique_ebtroot;
+
+    /***************************************************************************
+
+        Returns:
+            true if the map is empty or false if it contains elements.
+
+    ***************************************************************************/
+
+    public bool is_empty ( ) // const
+    {
+        return this.root.is_empty();
+    }
+
+    /***************************************************************************
+
+        Reinitialises the internal tree struct.
+
+        This method should be called only if it is known that the tree is empty,
+        otherwise pointers to malloc-allocated nodes are lost so they cannot be
+        deallocated.
+
+    ***************************************************************************/
+
+    public void reinit ( )
+    {
+        this.root = eb_root.init;
+        this.root.unique = true;
+    }
+
+    /***************************************************************************
+
+        It is possible to have an interface that gets only user elements passed,
+        not `eb64_node*` ebtree nodes. To use it, define the node item struct as
+        follows:
+         - Add a user element of some aggregate reference type (class or struct
+           pointer) to the node item struct and name it
+           `user_element_with_treemap_backlink`.
+         - To that class or struct add a public pointer to the node item struct
+           and name it `treemap_backlink`.
+
+        Example:
+        ---
+            class MyClass
+            {
+                struct MyTreeMapElement
+                {
+                    eb64_node ebnode; // mandatory
+                    MyClass user_element_with_treemap_backlink;
+                }
+
+                public MyTreeMapElement* treemap_backlink;
+            }
+
+
+            TreeMap!(MyTreeMapElement) map;
+
+            bool added;
+            MyClass obj  = map.put(12345, added, new MyClass);
+            MyClass obj2 = map.getUserElement(12345);
+            map.remove(obj);
+        ---
+
+    ***************************************************************************/
+
+    static if (is(typeof(Node.user_element_with_treemap_backlink) UserElement))
```

Comment:
Suggest removing this alternative API. It's confusing and doesn't seem to really add much.